doc/source/development/rfc/rfc103_schema_open_option.rst

```diff
@@ -0,0 +1,105 @@
+.. _rfc-103:
+
+===================================================================
+RFC 103 add a SCHEMA open option to selected OGR drivers
+===================================================================
+
+=============== =============================================
+Author:         Alessandro Pasotti
+Contact:        elpaso at itopen.it
+Status:         Draft
+Created:        2024-10-22
+=============== =============================================
+
+Summary
+-------
+
+This RFC enables users to specify a SCHEMA open option in the OGR
+drivers that support it.
+
+The new option will be used to override the auto-detected fields types.
+
+Motivation
+----------
+
+Several OGR drivers must guess the attribute data type: CSV, GeoJSON, SQLite,
+the auto-detected types are not always correct and the user has no way to
+override them at opening tim.
+
+For the details please see the discussion attached to the issue: https://github.com/OSGeo/gdal/issues/10943
+
+Implementation
+--------------
+
+A new open option named SCHEMA will be added to the following drivers:
+
+- CSV
+- GeoJSON
+- SQLite
+- GML
+
+The option will be used to specify the schema in the form of a JSON document (or a path to a JSON file).
+
+The structure of the JSON document will be similar to the one produced by the `ogrinfo -json` command
+with the notable exception that (for the scope of this RFC) only the type of the fields will be considered:
+
+.. code-block:: json
+    {
+    "fields": [
+        {
+        "name": "field1",
+        "type": "string"
+        },
+        {
+        "name": "field2",
+        "type": "integer"
+        }
+    ]
+    }
+
+
+In case of multi-layered datasets, the schema will be specified as a dictionary with the layer name as the key:
+
+.. code-block:: json
+    {
+    "layer1": {
+        "fields": [
+        {
+            "name": "field1",
+            "type": "string"
+        },
+        {
+            "name": "field2",
+            "type": "integer"
+        }
+        ]
+    },
+    "layer2": {
+        "fields": [
+        {
+            "name": "field1",
+            "type": "string"
+        },
+        {
+            "name": "field2",
+            "type": "integer"
+        }
+        ]
+    }
+    }
+
+A preliminary draft of the implementation can be found at:
+https://github.com/elpaso/gdal/commits/enhancement-gh10943-fields-schema-override/
+
+
+Errors and warnings
+-------------------
+
+- If the schema is not a valid JSON document, a critical error will be raised.
+
+- If the schema is a valid JSON document but does not contain the expected fields or it is a no-op
+  (does not contain any actionable instruction), a warning will be raised and the schema will be ignored.
+
+- Additional JSON properties will be ignored while parsing the schema.
+
+- If the schema contains a field that is not present in the dataset, a warning will be raised and the field will be ignored.
```

Comment:
What happens if some fields are missing?

Reply:
Nothing: auto-detected types will be used. The idea is that you can override only a part of the schema.

Reply:
I have added a sentence to clarify that partial overrides are possible (in fact, they probably are the main use case).

Reply:
I believe that it does not take long before someone wants to have an option to rename the fields by the same.
```
{
            "name": "field1",
            "new_name": "description"
            "type": "string"
 },
```

Reply:
And then someone would like to have an alternative for the OGR SQL EXCLUDE for removing some fields

> The syntax * EXCLUDE ([fields]) can be used to select all fields except those listed in parentheses.

They will not have that option if the fields which are missing from the schema are autodetected. But it is not possible to make both you and them happy at the same time, and you are the author...

Reply:
I believe we should add a required top-level JSON property "schema_type" whose only supported value currently will be "patch", to mean that we alter the autodetected schema for the parts we specify with the JSON document. If in the future we would want to allow full schema replacement, that would be done with "schema_type" = "full" or something like that

Reply:
@jratike80 I agree that field renaming will be very useful. Instead of an implicit operation, where `new_name` means replace `name` or `new_type` means replace `type`, I'd prefer to use an explicit operation (sort of like https://jsonpatch.com/#replace) which says: replace field `name` with `{"name": "NEW_NAME", "type": "NEW_TYPE"}`.

Reply:
Now that I think about it, field renaming can be done with SQL, right? While overriding the original field type cannot.

Reply:
SQL is flexible and type casting is supported. I seem to have used this kind of SQL with ogr2ogr
```
SELECT 
'FIN' AS "country",
CAST("mtk_id" as text) as "id", 
'CC BY 4.0' AS "license",
ST_Simplify("geometry",5) AS "geometry"
FROM "source"
```
I am not sure how well GDAL detects those CASTed attributes generally, but they have worked for my use case with  GeoPackage as the outputformat.

However, some users feel that a need to know some SQL is weird, and having several SQL dialects, from which GDAL makes a selection by hidden rules, makes more confusion for beginners. But it is hard to set the limits when to use SQL, when to add a new specific open option or ogr2ogr switch, or when to write an OGR VRT file.

For me the OGR_SCHEMA feels more user friendly than SQL CAST especially because I fear that different dialects may not do exactly similar casts. By supporting also renaming users would not need to to one part of mapping with OGR_SCHEMA and another one with SQL.

Reply:
> For me the OGR_SCHEMA feels more user friendly than SQL CAST

not only that, but type casting at the SQL level is too late. If a driver has wrongly guessed that a field was Integer32 wheras later records hold 64-bit values, the driver will present truncated/wrong values to the SQL engine. There's no way to "fix" that at that level. That must be something done by the driver itself